mmgen/models/architectures/biggan/modules.py

```diff
@@ -0,0 +1,200 @@
+        self.bn2 = BigGANConditionBatchNorm(
+            out_channels,
+            dim_after_concat,
+            shared_embedding=shared_embedding,
+            with_spectral_norm=with_spectral_norm,
+            auto_sync_bn=auto_sync_bn)
+
+        self.conv1 = ConvModule(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            conv_cfg=conv_cfg,
+            act_cfg=None,
+            with_spectral_norm=with_spectral_norm)
+
+        self.conv2 = ConvModule(
+            in_channels=out_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            conv_cfg=conv_cfg,
+            act_cfg=None,
+            with_spectral_norm=with_spectral_norm)
+
+    def forward(self, x, y):
+        x0 = self.bn1(x, y)
+        x0 = self.activation(x0)
+        if self.with_upsample:
+            x0 = self.upsample_layer(x0)
+            x = self.upsample_layer(x)
+        x0 = self.conv1(x0)
+        x0 = self.bn2(x0, y)
+        x0 = self.activation(x0)
+        x0 = self.conv2(x0)
+        if self.learnable_sc:
+            x = self.shortcut(x)
+        return x0 + x
+
+
+@MODULES.register_module()
+class BigGANConditionBatchNorm(nn.Module):
+
+    def __init__(self,
+                 num_features,
+                 input_dim,
+                 bn_eps=1e-5,
+                 sn_eps=1e-6,
+                 momentum=0.1,
+                 shared_embedding=True,
+                 with_spectral_norm=True,
+                 auto_sync_bn=True):
+        super(BigGANConditionBatchNorm, self).__init__()
+        assert num_features > 0 and input_dim > 0
+        # Prepare gain and bias layers
+        if shared_embedding:
+            self.gain = nn.Linear(input_dim, num_features, bias=False)
+            self.bias = nn.Linear(input_dim, num_features, bias=False)
+        else:
+            self.gain = nn.Embedding(input_dim, num_features)
+            self.bias = nn.Embedding(input_dim, num_features)
+
+        # please pay attention if shared_embedding is False
+        if with_spectral_norm:
+            self.gain = spectral_norm(self.gain, eps=sn_eps)
+            self.bias = spectral_norm(self.bias, eps=sn_eps)
+
+        self.bn = nn.BatchNorm2d(
+            num_features, eps=bn_eps, momentum=momentum, affine=False)
+
+        if auto_sync_bn and dist.is_initialized():
+            self.bn = SyncBatchNorm.convert_sync_batchnorm(self.bn)
+
+    def forward(self, x, y):
+        # Calculate class-conditional gains and biases
+        gain = (1 + self.gain(y)).view(y.size(0), -1, 1, 1)
+        bias = self.bias(y).view(y.size(0), -1, 1, 1)
+        out = self.bn(x)
+        return out * gain + bias
+
+
+@MODULES.register_module()
+class SelfAttentionBlock(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 conv_cfg=dict(type='Conv2d'),
```

Comment:
This `conv_cfg` is only for the `conv_cfg` in `ConvModule`. However, we hardly change this in this `ConvModule`. Because we only adopt the standard convolution here. Thus, I may recommend removing this config. Otherwise, you should use conv_cfg like:

```python
self.theta = ConvModule(
    self.in_channels,
    self.in_channels // 8,
    kernel_size=1,
    bias=False,
    act_cfg=None,
    padding=0,
    **conv_cfg)
```